twister: harness: Fix several issues at Console harness

golowanow opened the pull request:
Twister Console Harness several fixes and small improvements:

* Fix **unordered** pattern matching to treat the ztest as failed when not all of the expected patterns were found in the console output, but the ztest application itself reports 'PROJECT EXECUTION SUCCESSFUL'.
Extends #63621

* Unify debug logging on pattern match for ordered and unordered patterns.

* Check for non-empty pattern list configured for Console: If Console Harness has no regexp pattens configured, then ConfigurationError is raised, handled, and the failure of the test case correctly accounted in the summary results.

* Implement a workaround for Console harness to compose TestCase identifier correctly when Ztest suite with a single TestCase is executed (generally, Ztest suite should use Ztest harness).
Without the fix each TestCase result at Console is duplicated (and written into twister.json) with short 'identifier' set to TestSuite id only, without TestCase suffix added. The resulting entry with the full TestCase id is also stored, but its values are set empty to defaults.

golowanow committed:
twister: harness: Fix Console unordered pattern matching for ztest

Fix the Twister Console harness unordered pattern matching
to treat the ztest as failed when not all of the
expected patterns were found in the console output, but the ztest
application itself reports 'PROJECT EXECUTION SUCCESSFUL'.

Unify debug logging on pattern match for ordered and unordered patterns.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

gchwier commented on scripts/pylib/twister/twisterlib/harness.py:
```diff
@@ -218,16 +227,22 @@ def handle(self, line):
         if self.state == "passed" and self.ordered and self.next_pattern < self.patterns_expected:
-            logger.error(f"HARNESS:{self.__class__.__name__}: failed with only {self.next_pattern} matched patterns from expected {self.patterns_expected}")
+            logger.error(f"HARNESS:{self.__class__.__name__}: failed with"
+                         f" {self.next_pattern} of {self.patterns_expected}"
```
`{self.next_pattern+1}`

golowanow replied:
> `{self.next_pattern+1}`

why ? 
for `debug.coredump.backends.logging` it logs `ERROR   - HARNESS:Console: failed with 2 of 4 expected ordered patterns. `
for:
```
regex:                                                                                                                                                                                                                                       
  - "E: #CD:BEGIN#"                                                                                                                                                                                                                            
  - "E: #CD:5([aA])45([0-9a-fA-F]+)"                                                                                                                                                                                                          
  - "FOOO"
  - "E: #CD:END#"
```

gchwier replied:
Ok, I see it now. It is ordered pattern, so it receives number of matched patterns before failing one. First I thought that should be `3 of 4` in that case.

golowanow committed:
twister: harness: Check Console harness configuration

If Console Harness 'harness_config' properties have no 'regex'
patterns or no correct 'type' set, then ConfigurationError exception
is raised, handled, and the test instance error is accounted
in the summary results.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
twister: harness: Fix TestCase id at Console for Ztest

Implement a workaround for Console harness to compose TestCase
identifier correctly when a Ztest suite with a single testcase
uses this harness type. Normally, a Ztest suite should use the
Ztest Twister harness.
Without this workaround each Ztest TestCase result on Console is
duplicated (and written into twister.json) with its 'identifier'
attribute set to TestSuite id only, no TestCase suffix added;
the resulting entry with the full TestCase id is also stored,
but its values are empty with defaults.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow commented on scripts/pylib/twister/twisterlib/harness.py:
```diff
@@ -162,6 +162,9 @@ class Console(Harness):
         super(Console, self).configure(instance)
+        assert (
+            self.regex is not None and len(self.regex) > 0
+        ), f"HARNESS:{self.__class__.__name__}:no regex patterns configured"
```
done

gchwier commented on scripts/pylib/twister/twisterlib/harness.py:
```diff
@@ -162,6 +163,16 @@ class Console(Harness):
         super(Console, self).configure(instance)
+        if self.regex is None or len(self.regex) == 0:
+            self.state = "failed"
+            # Console Harness id has only TestSuite id without TestCase name suffix.
```
I see that getting tc_name is copied from line 256 (in handle method), there the comment is with `TODO`, that suggests something should be changed / solved later. Maybe it could be worth to extract it to separate method.

golowanow replied:
changed to a method, thank you.

gchwier reviewed (dismissed):
Looks good, one minor comment

golowanow commented:
@nashif, @galak, @gopiotr, @hakehuang, @PerMac - up to your attention

nashif commented:
> * Implement a workaround for Console harness to compose TestCase identifier correctly when only one TestCase is recognized as executed.
>   Without the fix each TestCase result at Console was duplicated (and written into twister.json) with short 'identifier' set to TestSuite id only, without TestCase suffix added.
>   The resulting entry with the full TestCase id was also stored, but without its values set being empty with defaults.

can you point to an example where this happens?

nashif commented on scripts/pylib/twister/twisterlib/harness.py:
```diff
@@ -162,6 +163,14 @@ class Console(Harness):
     def configure(self, instance):
         super(Console, self).configure(instance)
+        if self.regex is None or len(self.regex) == 0:
+            self.state = "failed"
```
if the definition of the testcase is wrong, this should be treated as an error. So, instead of doing this here and calling this a fail, error already during the parsing of the test (in config_parser.py)

golowanow replied:
> i wonder if this can be dealt with on the schema level, i.e.. disallowing empty sequences.

pykwalify has `allowempty` but for `map' type only
https://pykwalify.readthedocs.io/en/unstable/validation-rules.html#allowempty

golowanow replied:
> if the definition of the testcase is wrong, this should be treated as an error.

changed to 'error' status for instances

>  So, instead of doing this here and calling this a fail, error already during the parsing of the test (in config_parser.py)

it seems not easy to avoid these checks here with the current twister harness design: `testsute-schema.yaml` defines common `harness_config` for all types of harnesses simultaneously and parser checks for its presence with allowed (not required) properties. Later harness `configure()` needs to check what it got from the test instance.

golowanow commented:
> > current main branch (`92045a02`) with `./scripts/twister -vv -j 1 -p qemu_x86 -T tests/subsys/debug/coredump_backends`
> 
> ok. While your workaround might deal with the issue, the problem here is that the test is wrong. It can't be both a ztest and a console test. We either use ztest and rely on passing tests reporting based on ztest output or we rely on the console output, having both is wrong.

@nashif , that particular testcase wrong implementation (coredump_backends.loggign) is now addressed by #64937

hakehuang commented on scripts/pylib/twister/twisterlib/harness.py:
```diff
@@ -160,8 +161,26 @@ def run_robot_test(self, command, handler):
+        Console Harness id has only TestSuite id without TestCase name suffix.
+        TODO: Only the first TestCase name is taken if available.
+        '''
+        if self.instance is not None and len(self.instance.testcases) == 1:
```
in python shall we just use below?
```suggestion
        if self.instance and self.instance.testcases:
```

golowanow replied:
this workaround is applicable with a single Ztest testcase only.
I've simplified this `if` statement; also better description is added for this method as well as for the commit message.

hakehuang commented on scripts/pylib/twister/twisterlib/harness.py:
```diff
@@ -170,17 +189,29 @@ def configure(self, instance):
+        else:
+            self.state = "failed"
+            tc = self.instance.set_case_status_by_name(
+                self.id,
```
why not use self.get_testcase_name() for this?

golowanow replied:
good catch, I've forgot it here.

golowanow commented:
@nashif, @galak, @gchwier, @gopiotr, @hakehuang, @PerMac - up to your attention
